app/src/test/java/org/mozilla/fenix/components/metrics/AppStartupTelemetryTest.kt

```diff
@@ -0,0 +1,266 @@
+
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(UNKNOWN, WARM, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `GIVEN application exists and is backgrounded WHEN application started again through app icon but HomeActivity is recreated from savedInstanceState THEN records the correct values`() {
+        setupIntentMock(APP_ICON)
+        launchApplicationAndPutApplicationInBackground()
+
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), true)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(APP_ICON, WARM, true)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    private fun launchApplicationAndPutApplicationInBackground() {
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        // have to clear the mock function calls so it doesnt interfere with tests
+        clearMocks(metricController, answers = false)
+
+        appStartupTelemetry.onApplicationOnStop()
+    }
+
+    @Test
+    fun `GIVEN application is in background WHEN application is launched again HomeActivity only calls onResume THEN records the correct values`() {
+        setupIntentMock(UNKNOWN)
+        launchApplicationAndPutApplicationInBackground()
+
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(UNKNOWN, ERROR)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `GIVEN application is in background WHEN application is launched again HomeActivity calls onRestart but not onNewIntent THEN records the correct values`() {
+        setupIntentMock(APP_ICON)
+        launchApplicationAndPutApplicationInBackground()
+
+        appStartupTelemetry.onHomeActivityOnRestart()
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(UNKNOWN, HOT)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `GIVEN application is in background WHEN application is launched again and HomeActivity calls onNewIntent but not onRestart THEN records the correct values`() {
+        setupIntentMock(APP_ICON)
+        launchApplicationAndPutApplicationInBackground()
+
+        appStartupTelemetry.onHomeActivityOnNewIntent(intent.toSafeIntent())
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(APP_ICON, ERROR)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `WHEN AppAllStartup does not have savedInstanceState THEN do not return savedInstanceState`() {
```

Comment:
These `AppAllStartup` tests should technically be in the corresponding test file for where `AppAllStartup` is defined.

However, let's avoid another back and forth: let's land it as is and if you have time and you're feeling inspired afterwards, you can move it. :) Thanks for writing tests for these cases.

Reply:
The events have tons of data classes so would it make sense to create a single Events test file and put all the tests for all the classes there? or make a folder called event, and put all the individual files there? 

I think with a single file, `before/after `might get confusing? unless we use Junit5? which has nested support.